File: q20-q29/q22/q22.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        solution = []
        def backtrack(s = "", left = 0, right = 0):
            if len(s) == 2 * n:
                solution.append(s)
                return
            if left < n:
                backtrack(s+'(', left+1, right)
            if right < left:
                backtrack(s+')', left, right+1)
        
        backtrack()
        return solution

File: Sequence/q1-q99/q20-q29/q22/q22.cpp
#include <vector>
class Solution {
public:
    vector<string> solution;
    vector<string> generateParenthesis(int n) {
        string s = "";
        int left = 0, right = 0;
        backtrack(s, left, right, n);
        return this->solution;
    }
    void backtrack(string s, int left, int right, int n){
        if(s.size() == 2 * n){
            this->solution.push_back(s);
            return;
        }
        if (left < n){
            backtrack(s+"(", left+1, right, n);
        }
        if (right < left){
            backtrack(s+")", left, right+1, n);
        }
        
    }
};
